core/jvm/src/main/scala/fs2/interop/flow/package.scala

```diff
@@ -0,0 +1,118 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2
+package interop
+
+import cats.effect.kernel.{Async, Resource}
+import java.util.concurrent.Flow.Publisher
+
+/** Implementation of the reactive-streams protocol for fs2; based on Java Flow.
+  *
+  * @example {{{
+  * scala> import fs2.Stream
+  * scala> import fs2.interop.flow._
+  * scala> import java.util.concurrent.Flow.Publisher
+  * scala> import cats.effect.{IO, Resource}
+  * scala> import cats.effect.unsafe.implicits.global
+  * scala>
+  * scala> val upstream: Stream[IO, Int] = Stream(1, 2, 3).covary[IO]
+  * scala> val publisher: Resource[IO, Publisher[Int]] = upstream.toPublisher
+  * scala> val downstream: Stream[IO, Int] = Stream.resource(publisher).flatMap(p => p.toStream[IO](bufferSize = 16))
+  * scala>
+  * scala> downstream.compile.toVector.unsafeRunSync()
+  * res0: Vector[Int] = Vector(1, 2, 3)
+  * }}}
+  *
+  * @see [[http://www.reactive-streams.org/]]
+  */
+package object flow {
+
+  /** Creates a [[Stream]] from an [[Publisher]].
+    *
+    * The publisher only receives a subscriber when the stream is run.
+    *
+    * @param publisher The [[Publisher]] to transform.
+    * @param bufferSize setup the number of elements asked each time from the [[Publisher]].
+    *                   A high number can be useful if the publisher is triggering from IO,
+    *                   like requesting elements from a database.
+    *                   The publisher can use this `bufferSize` to query elements in batch.
+    *                   A high number will also lead to more elements in memory.
+    */
+  def fromPublisher[F[_], A](
+      publisher: Publisher[A],
+      bufferSize: Int
+  )(implicit
+      F: Async[F]
+  ): Stream[F, A] =
+    Stream
+      .eval(StreamSubscriber[F, A](bufferSize))
+      .flatMap { subscriber =>
+        subscriber.stream(
+          F.delay(publisher.subscribe(subscriber))
+        )
+      }
+
+  implicit final class PublisherOps[A](private val publisher: Publisher[A]) extends AnyVal {
+
+    /** Creates a [[Stream]] from an [[Publisher]].
+      *
+      * @param bufferSize setup the number of elements asked each time from the [[Publisher]].
+      *                   A high number can be useful is the publisher is triggering from IO,
+      *                   like requesting elements from a database.
+      *                   The publisher can use this `bufferSize` to query elements in batch.
+      *                   A high number will also lead to more elements in memory.
+      */
+    def toStream[F[_]](bufferSize: Int)(implicit F: Async[F]): Stream[F, A] =
```

Comment:
I am a bit skeptical if we should expose this, for the same reasons we don't expose a `toIO` syntax for `Future`. Specifically, the signature does not ask for a raw `Future`, but a suspended one:
```scala
object IO {
  def fromFuture[A](fut: IO[Future[A]]): IO[A]
}
```

Reply:
Well, AFAIK a `Publisher` is just a description, you can create one without needing to run any side effects.